Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp

```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
     element.clearAfterPseudoElement();
 }
 
+void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(RenderElement& renderer)
+{
+    auto destroyWritingSuggestionsIfNeeded = [&renderer, this]() {
+        if (!renderer.element())
+            return;
+
+        auto& editor = renderer.element()->document().editor();
+
+        if (WeakPtr writingSuggestionsRenderer = editor.writingSuggestionRenderer())
+            m_updater.m_builder.destroy(*writingSuggestionsRenderer);
+    };
+
+    if (!renderer.canHaveChildren())
+        return;
+
+    if (!renderer.element())
+        return;
+
+    auto& editor = renderer.element()->document().editor();
+
+    auto isWritingSuggestionElement = renderer.element() == editor.writingSuggestionsContainerElement();
+    if (!isWritingSuggestionElement)
+        return;
+
+    auto* writingSuggestionData = editor.writingSuggestionData();
+    if (!writingSuggestionData) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto style = renderer.getCachedPseudoStyle({ PseudoId::InternalWritingSuggestions }, &renderer.style());
+    if (!style || style->display() == DisplayType::None) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto* firstChildText = dynamicDowncast<RenderText>(renderer.firstChild());
+    if (!firstChildText) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    auto textWithoutSuggestion = firstChildText->text();
+
+    auto offset = writingSuggestionData->offset();
+    auto prefix = textWithoutSuggestion.substring(0, offset);
+    auto suffix = textWithoutSuggestion.substring(offset);
+
+    firstChildText->setText(prefix);
+
+    auto newStyle = RenderStyle::clone(*style);
+    if (auto writingSuggestionsRenderer = editor.writingSuggestionRenderer()) {
+        writingSuggestionsRenderer->setStyle(WTFMove(newStyle));
+
+        auto* writingSuggestionsText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->firstChild());
+        if (!writingSuggestionsText) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```

Comment:
```suggestion
        RELEASE_ASSERT(writingSuggestionsText);
```
I have a slight preference for something like this in terms of sanity.